src/field.rs

```diff
@@ -31,6 +44,28 @@ pub struct Fp31(<Self as Field>::Integer);
 impl Field for Fp31 {
     type Integer = u8;
     const PRIME: Self::Integer = 31;
+    const ZERO: Self = Fp31(0);
+    const ONE: Self = Fp31(1);
+
+    fn invert(&self) -> Self {
```

Comment:
that requires further constraining `Integer` type as we need to be able to subtract `2_u8` from `PRIME` and left shift to compute the `pow`.  I can either depend on `num-traits` to bound `Integer: PrimInt` or create a simple trait for it. I think the latter is slightly better as we don't need the full power of num-traits for now

Reply:
Yeah, I get that.

Would it be easier to make `type PRIME : Field` instead?  Then you could do `Self::PRIME - Self::from(2_u128)`.  Of course, then you have the problem that Self::PRIME isn't in the field...

Reply:
I have a workaround for `invert` that only requires `Field: Into<Self::Integer>` and `type Integer: Ord + Sub<Output = Self::Integer>` bounds

```rust
 /// computes the multiplicative inverse of `self`. It is UB if `self` is 0.
    #[must_use]
    fn invert(&self) -> Self {
        debug_assert!(
            self != &Self::ZERO,
            "Multiplicative inverse is not defined for 0"
        );

        self.pow(Self::PRIME - Self::ONE.into() - Self::ONE.into())
    }
```

for `pow` is a little more tricky and making `Prime: Field` is not helpful because we need right shift for `Self::Integer`:

```rust
pub trait Int :
    Sized
    + Copy
    + Debug
    + Ord
    + Sub<Output = Self>
    + Into<u128>
    + Shr<u32, Output = Self>
    + BitAnd<u8, Output = Self>
    + PartialEq
{
    const ZERO: Self;
    const BITS: u32;
}

impl Int for u8 {
    const ZERO: Self = 0_u8;
    const BITS: u32 = u8::BITS;
}

pub trait Field:
{
  type Integer: Int;
  ...
}
```

I don't see any issues here and I can raise a PR if this is acceptable

Reply:
Sgtm. You can implement a default From implementation for the new Int type as well.

Reply:
When implementing `From<Int>` for `Field` we would run into the same issue with dyn :(